webapp/src/cloud.ts

```diff
@@ -0,0 +1,298 @@
+                        // last write wins.
+                        if (local.modificationTime > remoteFile.header.modificationTime) {
+                            if (local.cloudVersion === remoteFile.version) {
+                                // local is one ahead, push as normal
+                                pxt.debug(`local project '${local.name}' has changes that will be pushed to the cloud.`)
+                            } else {
+                                // conflict and local wins
+                                // TODO: Pop a dialog and/or show the user a diff. Better yet, handle merges.
+                                pxt.log(conflictStr + ' Local will overwrite cloud.')
+                            }
+                            const text = await workspace.getTextAsync(local.id);
+                            const newVer = await setAsync(local, remoteFile.version, text);
+                            U.assert(!!newVer, 'Failed to sync local change (1)');
+                            // save to the workspace header again to make sure cloud metadata gets saved
+                            await workspace.saveAsync(local, null, true)
+                        } else {
+                            // conflict and remote wins
+                            // TODO: Pop a dialog and/or show the user a diff. Better yet, handle merges.
+                            pxt.log(conflictStr + ' Cloud will overwrite local.')
+                            const newHeader = {...local, ...remoteFile.header} // make sure we keep local-only metadata like _rev
+                            localHeaderChanges[remoteFile.header.id] = newHeader
+                            const localVer = await workspace.saveAsync(newHeader, remoteFile.text, true);
+                        }
+                    }
+                } else {
+                    if (local.isDeleted) {
+                        // Delete remote copy.
+                        //return deleteAsync(local, local.cloudVersion);
+                        // Mark remote copy as deleted.
+                        remote.isDeleted = true;
+                        const newVer = await setAsync(remote, local.cloudVersion, {});
+                        const newHeader = {...local, ...remote} // make sure we keep local-only metadata like _rev
+                        await workspace.saveAsync(newHeader, null, true);
+                        U.assert(!!newVer, 'Failed to sync local change (2)');
+                    }
+                    if (remote.isDeleted) {
+                        // Delete local copy.
+                        local.isDeleted = true;
+                        localHeaderChanges[local.id] = local
+                        return workspace.forceSaveAsync(local, {}, true)
+                            .then(() => { data.clearCache(); }) // TODO @darzu: is this cache clear necessary?
```

Comment:
@eanders-MS , do you remember why you added this?

Reply:
Are you asking about the call to clearCache?

Reply:
Yes